### ISSUE TITLE
Crash on simple transaction

### ISSUE THREAD

**Issue by bodnar-dan** [realm/realm-swift#7682]
### How frequently does the bug occur?

Sometimes

### Description

After updating from 10.20.0 to 10.22.0, this particular transaction (old code) is crashing for some users. 
```
let realm: RLMRealm = try .init(configuration: aRealmConfiguration)
object = getObject(in: realm)
try realm.transaction {
      object.lastOpen = Date()
}
...
```
Those several users are blocked and unable to use the app because of this issue. For them this is happening repetitively. We're calling this transaction at some point soon after the app is either launched or transitioned from background, but it's not the first Realm access (**although it's possible to be the first write on Realm**). The object is retrieved in the same realm and everything happens on main thread.
One thing to mention, we are using ObjC models, `lastOpen` is a non-nullable propert, but has a default value (we also have this object's property since the initial schema).

### Stacktrace & log output

```shell
Crashed: com.apple.main-thread
0  libsystem_kernel.dylib         0x7964 __pthread_kill + 8
1  libsystem_pthread.dylib        0x7378 pthread_kill + 268
2  libsystem_c.dylib              0x1ff50 abort + 164
3  Realm                          0x2f07fc please_report_this_issue_in_github_realm_realm_core + 10
4  Realm                          0x2f0afc realm::util::terminate_internal(std::__1::basic_stringstream<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) + 328
5  Realm                          0x2f0c74 realm::util::terminate(char const*, char const*, long, std::initializer_list<realm::util::Printable>&&) + 328
6  Realm                          0x5bc34 void realm::util::terminate<unsigned long, unsigned char>(char const*, char const*, long, unsigned long, unsigned char) + 46
7  Realm                          0x151310 realm::Array::set(unsigned long, long long) + 234
8  Realm                          0x2191bc realm::Node::do_copy_on_write(unsigned long) + 232
9  Realm                          0x1512a0 realm::Array::set(unsigned long, long long) + 124
10 Realm                          0x21ad9c realm::Obj& realm::Obj::set<realm::Timestamp>(realm::ColKey, realm::Timestamp, bool) + 468
11 Realm                          0x7dfc void (anonymous namespace)::kvoSetValue<NSDate* __strong, NSDate* __strong>(RLMObjectBase*, unsigned long, NSDate* __strong) + 144
12 Realm                          0x1299c ___ZN12_GLOBAL__N_110makeSetterIU8__strongP6NSDateS3_EEP11objc_objectP11RLMProperty_block_invoke_2 + 24
```

### Can you reproduce the bug?

Not yet

### Reproduction Steps

We are unable to reproduce, however the crash is visible on crashlytics reports.

### Version

10.22.0

### What SDK flavour are you using?

Local Database only

### Are you using encryption?

No, not using encryption

### Platform OS and version(s)

different iOS versions

### Build environment

Xcode version: 13.2.1, but not sure if it's relevant.
Dependency manager and version: Carthage

**Comment by pavel-ship-it** [comment]
Hi @Bodnar-Dan , thank you for the report.

Can you show `aRealmConfiguration`?
Sometimes you can reproduce issues on Release build.

**Comment by Bodnar-Dan** [comment]
Hello @pavel-ship-it 

We use ObjC for that. 
```
+ (instancetype)realmConfiguration {
    static dispatch_once_t onceRealmConfigurationToken;
    dispatch_once(&onceRealmConfigurationToken, ^{
        __block uint64_t newSchemaVersion = 200;
        path = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES).firstObject; 
        NSString *realmDirectoryPath = [path stringByAppendingPathComponent:@"realm"];
        NSString *realmFilePath = [realmDirectoryPath stringByAppendingPathComponent:@"app.realm"];
        
        RLMRealmConfiguration *configuration = [[RLMRealmConfiguration alloc] init];
        configuration.fileURL = [NSURL URLWithString:realmFilePath];
        configuration.schemaVersion = newSchemaVersion;
        configuration.objectClasses = @[
            ...
        ];
        configuration.migrationBlock = ^(RLMMigration *migration, uint64_t oldSchemaVersion) {
            ...
        };
        aStaticConfiguration = configuration;
        [RLMRealmConfiguration prepareRealmDirectoryAtPath:realmDirectoryPath]; // This only creates the directories (if not existing) with attributes `@{NSFileProtectionKey: NSFileProtectionNone}`
    });
    return aStaticConfiguration;
}
```

**Comment by pavel-ship-it** [comment]
Supposed to be fixed. I'll get back to that after release.

**Comment by tgoyne** [comment]
This might be fixed in 10.24.0. 10.22.0 pulled in some internal changes to how write transactions work which had some bugs. 10.24.0 fixes all the problems we were able to hit.

**Comment by Bodnar-Dan** [comment]
I am still seeing this with 10.24.0 :(
